File: zephyr/app/common/projection.c
#include <math.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#include "projection.h"

struct vec2 projection_project(struct vec3 point, int screen_w, int screen_h,
			       const struct camera3d *cam)
{
	const float f = 1.0f / tanf((cam->fov_deg * 0.5f) * ((float)M_PI / 180.0f));
	const float z = fmaxf(point.z, cam->near_plane);
	const float sx = (point.x * f / (z * cam->aspect)) * (screen_w * 0.5f) + screen_w * 0.5f;
	const float sy = (-point.y * f / z) * (screen_h * 0.5f) + screen_h * 0.5f;

	return (struct vec2){ sx, sy };
}

struct vec3 projection_unproject(struct vec2 screen, float depth, int screen_w, int screen_h,
				 const struct camera3d *cam)
{
	const float f = 1.0f / tanf((cam->fov_deg * 0.5f) * ((float)M_PI / 180.0f));
	const float z = fmaxf(depth, cam->near_plane);
	const float nx = (screen.x - screen_w * 0.5f) / (screen_w * 0.5f);
	const float ny = -(screen.y - screen_h * 0.5f) / (screen_h * 0.5f);

	return vec3_make(nx * z * cam->aspect / f, ny * z / f, z);
}

File: android/ESP32S3ImuSim/app/src/main/java/com/esp32s3/imusim/Projection.kt
package com.esp32s3.imusim

import kotlin.math.tan

data class Vec2(val x: Float, val y: Float)
data class Vec3(val x: Float, val y: Float, val z: Float)

data class Camera3D(
    val fovDeg: Float = 45f,
    val nearPlane: Float = 0.1f,
    val aspect: Float = 1f,
)

object Projection {
    fun project(point: Vec3, screenW: Int, screenH: Int, cam: Camera3D): Vec2 {
        var z = point.z
        if (z < cam.nearPlane) z = cam.nearPlane
        val f = 1f / tan(Math.toRadians((cam.fovDeg / 2f).toDouble())).toFloat()
        val sx = (point.x * f / (z * cam.aspect)) * (screenW / 2f) + (screenW / 2f)
        val sy = (-point.y * f / z) * (screenH / 2f) + (screenH / 2f)
        return Vec2(sx, sy)
    }

    fun unproject(screen: Vec2, depth: Float, screenW: Int, screenH: Int, cam: Camera3D): Vec3 {
        val f = 1f / tan(Math.toRadians((cam.fovDeg / 2f).toDouble())).toFloat()
        val nx = (screen.x - screenW / 2f) / (screenW / 2f)
        val ny = -(screen.y - screenH / 2f) / (screenH / 2f)
        val x = nx * depth * cam.aspect / f
        val y = ny * depth / f
        return Vec3(x, y, depth)
    }

    fun transform(m: FloatArray, v: Vec3): Vec3 {
        return Vec3(
            m[0] * v.x + m[1] * v.y + m[2] * v.z,
            m[3] * v.x + m[4] * v.y + m[5] * v.z,
            m[6] * v.x + m[7] * v.y + m[8] * v.z,
        )
    }

    fun applyZoom(zoom: Float, v: Vec3): Vec3 = Vec3(v.x * zoom, v.y * zoom, v.z * zoom)

    fun applyZoom(zx: Float, zy: Float, zz: Float, v: Vec3): Vec3 = Vec3(v.x * zx, v.y * zy, v.z * zz)
}
